source/common/upstream/cds_api_impl.cc

```diff
@@ -59,8 +63,19 @@ void CdsApiImpl::onConfigUpdate(const std::vector<Config::DecodedResourceRef>& r
 void CdsApiImpl::onConfigUpdate(const std::vector<Config::DecodedResourceRef>& added_resources,
                                 const Protobuf::RepeatedPtrField<std::string>& removed_resources,
                                 const std::string& system_version_info) {
-  auto exception_msgs =
-      helper_.onConfigUpdate(added_resources, removed_resources, system_version_info);
+  std::vector<std::string> exception_msgs;
+  // Run the external config validators (if any).
+  TRY_ASSERT_MAIN_THREAD { executeValidators(added_resources, removed_resources); }
```

Comment:
I can't remember, but did we discuss doing this in `GrpcSubscriptionImpl::onConfigUpdate`?

Reply:
Yes, as we discussed offline, this should be moved to the `onConfigUpdate` that is being called on the singe update message containing all the resources of the same type.